packages/babel-parser/src/parse-error.js

```diff
@@ -0,0 +1,211 @@
+  }
+}
+
+function toParseErrorClass<ErrorDetails>(
+  toMessage: ToMessage<ErrorDetails>,
+  credentials: ParseErrorCredentials,
+): Class<ParseError<ErrorDetails>> {
+  return class extends ParseError<ErrorDetails> {
+    #message: typeof StandardMessage | string = StandardMessage;
+
+    constructor(...args): ParseError<ErrorDetails> {
+      super(...args);
+      // $FlowIgnore - Only necessary because we can't make syntaxPlugin optional.
+      ObjectAssign(this, credentials);
+      return this;
+    }
+
+    get message() {
+      return this.#message !== StandardMessage
+        ? String(this.#message)
+        : `${toMessage(this.details)} (${this.loc.line}:${this.loc.column})`;
+    }
+
+    set message(message) {
+      this.#message = message;
+    }
+  };
+}
+
+// This part is tricky, and only necessary due to some bugs in Flow that won't
+// be fixed for a year(?): https://github.com/facebook/flow/issues/8838
+// Flow has a very difficult time extracting the parameter types of functions,
+// so we are forced to pretend the class exists earlier than it does.
+// `toParseErrorCredentials` *does not* actuall return a
+// `Class<ParseError<ErrorDetails>>`, but we simply mark it as such to "carry"
+// the `ErrorDetails` type parameter around. This is not a problem in Typescript
+// where this intermediate function actually won't be needed at all.
+declare function toParseErrorCredentials<T: string>(
+  T,
+  ?{ code?: ParseErrorCode, reasonCode?: string } | boolean,
+): Class<ParseError<{||}>>;
+
+// ESLint seems to erroneously think that Flow's overloading syntax is an
+// accidental redeclaration of the function:
+// https://github.com/babel/eslint-plugin-babel/issues/162
+// eslint-disable-next-line no-redeclare
+declare function toParseErrorCredentials<T>(
+  (T) => string,
+  ?{ code?: ParseErrorCode, reasonCode?: string } | boolean,
+): Class<ParseError<T>>;
+
+// See comment about eslint and Flow overloading above.
+// eslint-disable-next-line no-redeclare
+export function toParseErrorCredentials(toMessageOrMessage, credentials) {
+  return [
+    typeof toMessageOrMessage === "string"
+      ? () => toMessageOrMessage
+      : toMessageOrMessage,
+    credentials,
+  ];
+}
+
+declare function toParseErrorClasses(string[]): typeof toParseErrorClasses;
+
+// See comment about eslint and Flow overloading above.
+// eslint-disable-next-line no-redeclare
+declare function toParseErrorClasses<T: Object>(
+  toClasses: (typeof toParseErrorCredentials) => T,
+  syntaxPlugin?: string,
+): T;
+
+// See comment about eslint and Flow overloading above.
+// eslint-disable-next-line no-redeclare
+export function toParseErrorClasses(argument, syntaxPlugin) {
+  if (ArrayIsArray(argument)) {
+    return toClasses => toParseErrorClasses(toClasses, argument[0]);
+  }
+
+  const classes = argument(toParseErrorCredentials);
+
+  for (const reasonCode of ObjectKeys(classes)) {
+    const [toMessage, credentials = {}] = classes[reasonCode];
```

Comment:
If we need both the key and the value, we can directly iterate over `Object.entries(classes)`.

Reply:
I use `Object.keys` here because when I use `Object.entries` was added in Node 8. I wasn't sure whether our compilation process did any fancy polyfilling, but I'm pretty sure the Node 6 tests failed when I used `Object.entries`.